### ISSUE TITLE
training a language model from scratch in tensorflow?

### ISSUE THREAD

**Issue by randomgambit** [huggingface/blog#137]
Good morning/evening?

Thanks for the amazing blog posts and the amazing `huggingface` package!

I am using `tensorflow` and I realized the nice colab notebook on training a language model from scratch is only available in `pytorch`... https://github.com/huggingface/blog/blob/master/notebooks/01_how_to_train.ipynb

Is there a `tensorflow` version somewhere? That would be amazing!

Thanks!

**Comment by Rocketknight1** [comment]
Hi, Tensorflow maintainer here! This is something I'm working on, but it may be a couple of weeks until the TF versions of the notebooks are ready. In the meantime, you can also check out our examples [here](https://github.com/huggingface/transformers/tree/master/examples/tensorflow/language-modeling), though note that these are a bit longer and more complex, since they include support for various special cases, as well as things like multi-GPU and TPU training.

**Comment by randomgambit** [comment]
wow!!! just a couple weeks??? a dream comes true! Happy to test your preliminary notebook if you want! I have a quite powerful GPU. Thanks @Rocketknight1

**Comment by randomgambit** [comment]
@Rocketknight1 happy to work even with incomplete notebooks! :)

**Comment by randomgambit** [comment]
@Rocketknight1 I hope all is well! Just touching base on the TF2 notebook. Is there anything I can do to help you wrap this up? Happy to test anything! A Tensorflow notebook for language modeling would be amazing! What do you think?
Thanks!

**Comment by Rocketknight1** [comment]
Hey! Don't worry, it's still coming very soon on my roadmap. Right now I'm working on a revamp of our `DataCollator` classes and our `Dataset` classes. This will enable us to easily convert data to `tf.data.Dataset`. We want this to be our standard method for loading data for Tensorflow/Keras, so I need to finish this up before I make the Tensorflow ports of the notebooks, so I can use it there.

Right now I'm expecting to finish the data collators probably tomorrow, but there'll definitely be a period of review and bugfixing for them and the changes to `Dataset` before they go live. While that's happening, I'll start working on the notebooks. Hopefully I'll have something to show by next week, and I'll let you know whenever I push a language modelling notebook.

**Comment by randomgambit** [comment]
Thank you! This is great!! Let me know if you need any kind of feedback. Thanks!

**Comment by Rocketknight1** [comment]
Hi @randomgambit ,

That's not the one I'm working on. I've just pushed a beta version of the new notebook for language modeling with TF [here](https://github.com/huggingface/notebooks/blob/new_tf_notebooks/examples/tf_language_modeling.ipynb). Note that it will not work unless you install both the most recent version of `transformers` with `pip`, and my branch of `datasets` [here](https://github.com/huggingface/datasets/tree/tf_dataset_conversion), which you'll have to download manually (make sure you get the `tf_dataset_conversion` branch!) and then `pip install .`

Over the next couple of weeks we're going to be merging these branches to master and making proper releases that support these new features, but you're stuck with the messy solution for now. If you try the notebook anyway, let me know how it goes!

**Comment by randomgambit** [comment]
thanks @Rocketknight1 this is great! I am looking at it right now. Just a (perhaps) naive question. Given that we are using keras at the end and transforming the data to a tf.dataset, cant we simply process the data inside a pandas dataframe instead of a huggingface dataset? That will solve the issue of having to modify the datasets branch? Am I missing something? Thanks!!

**Comment by randomgambit** [comment]
Actually I wrote the lines below which seem to work fine (still running on my colab)

```
train_set = (
    tf.data.Dataset.from_tensor_slices(
        (
            tf.cast(lm_datasets['train']['attention_mask'], tf.int32),
            tf.cast(lm_datasets['train']['input_ids'], tf.int32),
            tf.cast(lm_datasets['train']['labels'], tf.int32)
        )
    )
)


validation_set = (
    tf.data.Dataset.from_tensor_slices(
        (
            tf.cast(lm_datasets['validation']['attention_mask'], tf.int32),
            tf.cast(lm_datasets['validation']['input_ids'], tf.int32),
            tf.cast(lm_datasets['validation']['labels'], tf.int32)
        )
    )
)

model.fit(train_set, validation_data=validation_set, epochs=1)
```

I imagine this is more naive and simple than using the function to wrote to convert from `datasets` to `tf.dataset`? Is that right?

Thanks!

**Comment by randomgambit** [comment]
actually.... I obtain a perplexity of zero... which is not good... what do you get instead? There is no output in the colab you shared.

Thanks!

**Comment by randomgambit** [comment]
Hi @Rocketknight1 I dont know if you're around these days, but please let me know which perplexity you get so that I can compare with the pandas alternative I provided above. Or perhaps I am completely misguided here? Thanks!!! what you do is really awesome.

**Comment by Rocketknight1** [comment]
I don't follow exactly - how would you handle this with Pandas?

**Comment by randomgambit** [comment]
hi @Rocketknight1, please see the fully reproducible code below and let me know what you think (note, I use a pandas dataframe but I quickly convert to a simple `datasets` instance. The point is that we could perhaps convert to a `tf.dataset` by using the native tensorflow function `from_tensor_slices`? What do you think of this code?

Thanks!!

```


from datasets import load_dataset
datasets = load_dataset('wikitext', 'wikitext-2-raw-v1')

from transformers import AutoTokenizer
import pandas as pd
import numpy as np
import datasets
import transformers
import tensorflow

from datasets import load_dataset
mydata = load_dataset('wikitext', 'wikitext-2-raw-v1')

#lets make up a pandas dataframe from this data ... just for the example

df_train = pd.DataFrame({'text' : mydata['train']['text']})
df_test = pd.DataFrame({'text' : mydata['test']['text']})

model_checkpoint = "gpt2"
tokenizer_checkpoint = "sgugger/gpt2-like-tokenizer"
    
tokenizer = AutoTokenizer.from_pretrained(tokenizer_checkpoint)

dataset_train = datasets.Dataset.from_pandas(df_train[['text']])
dataset_test = datasets.Dataset.from_pandas(df_test[['text']])

def tokenize_function(mydf):
    return tokenizer(mydf["text"])

tokenized_train = dataset_train.map(tokenize_function, batched=True, remove_columns = ['text'])
tokenized_test = dataset_test.map(tokenize_function, batched=True, remove_columns = ['text'])


block_size = 128

def group_texts(examples):
    # Concatenate all texts.
    concatenated_examples = {k: sum(examples[k], []) for k in examples.keys()}
    total_length = len(concatenated_examples[list(examples.keys())[0]])
    # We drop the small remainder, we could add padding if the model supported it instead of this drop, you can
        # customize this part to your needs.
    total_length = (total_length // block_size) * block_size
    # Split by chunks of max_len.
    result = {
        k: [t[i : i + block_size] for i in range(0, total_length, block_size)]
        for k, t in concatenated_examples.items()
    }
    result["labels"] = result["input_ids"].copy()
    return result


lm_train = tokenized_train.map(
    group_texts,
    batched=True,
    batch_size=1000
)

lm_test = tokenized_test.map(
    group_texts,
    batched=True,
    batch_size=1000
)


tokenizer.decode(lm_train[0]["input_ids"])

from transformers import TFAutoModelForCausalLM
model = TFAutoModelForCausalLM.from_pretrained(model_checkpoint)
import tensorflow as tf


# conversion by hand, less efficient (batch?) but should work?


train_set = (
    tf.data.Dataset.from_tensor_slices(
        (
            tf.cast(lm_train['attention_mask'], tf.int32),
            tf.cast(lm_train['input_ids'], tf.int32),
            tf.cast(lm_train['labels'], tf.int32)
        )
    )
)


test_set = (
    tf.data.Dataset.from_tensor_slices(
        (
            tf.cast(lm_test['attention_mask'], tf.int32),
            tf.cast(lm_test['input_ids'], tf.int32),
            tf.cast(lm_test['labels'], tf.int32)
        )
    )
)


from transformers import create_optimizer, AdamWeightDecay
optimizer = AdamWeightDecay(lr=2e-5,
    weight_decay_rate=0.01
)

import tensorflow as tf
def dummy_loss(y_true, y_pred):
    return tf.reduce_mean(y_pred)

model.compile(optimizer=optimizer, loss={'loss': dummy_loss})

model.fit(train_set, validation_data=test_set, epochs=1)


```

**Comment by randomgambit** [comment]
strangely enough this gives

```
import math

eval_loss = model.evaluate(test_set)[0]
print(f"Perplexity: {math.exp(eval_loss):.2f}")

2203/2203 [==============================] - 12s 6ms/step - loss: -945133.3750 - past_key_values_1_loss: -945133.3750
Perplexity: 0.00
```

**Comment by randomgambit** [comment]
@Rocketknight1 did you have a chance to have a look? happy to help!

**Comment by Rocketknight1** [comment]
Well, one thing I can tell you is that that loss is not a normal loss, lol. The reason is, I suspect, that when you made the dataset you lost the names of the input Tensors - so you have the `input_ids`, `attention_mask` and `labels` that you need, but they're in a random order. The first three arguments to `TFGPT2LMHeadModel` are `input_ids`, `past` and `attention_mask` so I suspect that's what the model thinks you're giving it. The `to_tf_dataset()` method handles this correctly, so you shouldn't get this problem with the official notebooks.

Your method to create a dataset should work, though, at the cost of loading the whole dataset into memory. All you need to do is add a function that converts the output tuples from that dataset to dicts with correct key names and training should go a lot more smoothly.

Also, sorry for the slow replies here! This isn't really a place for debugging user code, so I can only really get to it after I clear out the rest of my schedule.

**Comment by randomgambit** [comment]
Hi @Rocketknight1 ! It seems the method `to_tf_dataset` made it to the `master` version of `datasets`? Ready for prime show? Thanks!

**Comment by Rocketknight1** [comment]
Hey, just closing some old issues I'm still on - all the notebooks have been fully up-to-date for a while!
